### ISSUE TITLE
[CI] WatcherMappingUpdateIT testMappingsAreUpdated {cluster=UPGRADED} failing

### ISSUE THREAD

**Issue by thecoop** [elastic/elasticsearch#100282]
Looks like its related to the 8.12 bump

**Build scan:**
https://gradle-enterprise.elastic.co/s/2vnbfdltr7inm/tests/:x-pack:qa:full-cluster-restart:v8.11.0%23bwcTest/org.elasticsearch.xpack.restart.WatcherMappingUpdateIT/testMappingsAreUpdated%20%7Bcluster=UPGRADED%7D
**Reproduction line:**
```
./gradlew ':x-pack:qa:full-cluster-restart:v8.11.0#bwcTest' -Dtests.class="org.elasticsearch.xpack.restart.WatcherMappingUpdateIT" -Dtests.method="testMappingsAreUpdated {cluster=UPGRADED}" -Dtests.seed=133D25BE309D83ED -Dtests.bwc=true -Dtests.locale=ca -Dtests.timezone=Indian/Maldives -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.restart.WatcherMappingUpdateIT&tests.test=testMappingsAreUpdated%20%7Bcluster%3DUPGRADED%7D
**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: a string containing "\"version\":\"8.12.0\""
     but: was "{".watches":{"mappings":{"dynamic":"strict","_meta":{"managed_index_mappings_version":1,"version":"8.11.0"},"properties":{"actions":{"type":"object","dynamic":"true","enabled":false},"condition":{"type":"object","dynamic":"true","enabled":false},"input":{"type":"object","dynamic":"true","enabled":false},"metadata":{"type":"object","dynamic":"true"},"status":{"type":"object","dynamic":"true","enabled":false},"throttle_period":{"type":"keyword","index":false,"doc_values":false},"throttle_period_in_millis":{"type":"long","index":false,"doc_values":false},"transform":{"type":"object","dynamic":"true","enabled":false},"trigger":{"type":"object","dynamic":"true","enabled":false}}}}}"

  at __randomizedtesting.SeedInfo.seed([133D25BE309D83ED:7045505F1A71BDDB]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.xpack.restart.WatcherMappingUpdateIT.lambda$assertMappingVersion$0(WatcherMappingUpdateIT.java:82)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1196)
  at org.elasticsearch.xpack.restart.WatcherMappingUpdateIT.assertMappingVersion(WatcherMappingUpdateIT.java:77)
  at org.elasticsearch.xpack.restart.WatcherMappingUpdateIT.testMappingsAreUpdated(WatcherMappingUpdateIT.java:72)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster$1.evaluate(DefaultLocalElasticsearchCluster.java:39)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

**Comment by mark-vieira** [comment]
I've muted this for now.

**Comment by williamrandolph** [comment]
Looks systemindex-ish. I'll take a look.

**Comment by ldematte** [comment]
I think I understand what's happening.
The test is starting with a 8.11 cluster (the "old" cluster)
Here it creates a watch. In 8.11 we have already the specific metadata version in system indices:
```
"_meta":{"managed_index_mappings_version":1,"version":"8.11.0"}
```
Then, it proceeds to upgrade the cluster; on the new cluster the test reads the mappings, and it expects to find it "upgraded", with `"version":"8.12.0"`. 
However, https://github.com/elastic/elasticsearch/pull/99668 introduced the new upgrade logic based on `managed_index_mappings_version`, which ignores the cluster version. Since we did not change mappings between 8.11 and 8.12, no upgrade happened and the entire _meta remained the same.

Now, to me it seems this is working as intended, but I need @williamrandolph input here.
The `"version"` field will remain stuck to either the last _cluster_ version in which the mapping changed, or the version in which the new versioning schema was introduced (8.11), or the version with which the index was created. It will be eventually removed, and it will not be used for any decision beyond 8.10 (as versions from 8.11 will use the new field), so it's OK to remain behind (actually, it could even be hardcoded to 8.11, as I believe we did in other places).

If, as I suspect, this is not a problem but the desired behaviour, this tests will become trickier to write: upgrading the cluster will not necessarily trigger a mapping update anymore.
I suggest to remove the test, or skip it for > 8.11, as mapping update should already be covered by `SystemIndexMappingUpdateServiceIT` and `SystemIndexMappingUpdateServiceTests`.

Actually, I see there is a TODO there by @williamrandolph to add more tests for old/new mixed conditions, it might be worth doing it as a replacement for this test.

**Comment by williamrandolph** [comment]
I think we do have the general test we want in `SystemIndexMappingUpdateServiceIT#testSystemIndexManagerUpgradesMappings`, which uses a mock system index descriptor where we can swap mappings at test time.

If we wanted to test with a full cluster restart, I think we'd want to pull the minimum version for the watches index out of cluster state... but this style of test is less useful now that we can't strictly compare with Versions. For example, we can't directly use the `Watcher.WATCHES_INDEX_MAPPINGS_VERSION` value, so we'd have to do something complicated to figure out the expected current value.

I believe Watcher relies fully on the system index mappings management (unlike Security and ML, which update mappings more eagerly than the system index system does).

**Comment by ldematte** [comment]
Hello @williamrandolph 

I followed your findings, and I can confirm everything you wrote:
- `SystemIndexMappingUpdateServiceIT#testSystemIndexManagerUpgradesMappings` is covering the update part; instead of a version upgrade + cluster restart it uses an artificial change + status update, but the effects should be the same
- The stated purpose of `WatcherMappingUpdateIT#testMappingsAreUpdated` is (was) the same, but it will not work anymore (and rightly so) as a cluster upgrade does not guarantee a mappings update.
- It would be very difficult to write a meaningful and reliable check at cluster upgrade time: what could we assert? Sometime the expected metadata version will be the same, sometimes it will be a new version.
- Watcher seems indeed to use system index mappings management, so it would make sense to add tests to `SystemIndexMappingUpdateServiceIT`, if any

The only thing I could think of, is to restrict the test to pre-8.11 versions (with assumeTrue) to test that the metadata has been updated correctly to the new format, and/or splut a test for post-8.11 and test that the metadata version is equal or grater (but I think this will be a really lousy test :) )

Probably the most efficient solution is to just remove this test, as the mapping update concern is already covered by the test you linked.
